syntax = "proto3";

package cortical;

// Edge represents a typed connection with metadata
message Edge {
  string target_id = 1;
  double weight = 2;
  string relation_type = 3;
  double confidence = 4;
  string source = 5;
}

// Minicolumn represents a cortical minicolumn
message Minicolumn {
  string id = 1;
  string content = 2;
  int32 layer = 3;
  double activation = 4;
  int32 occurrence_count = 5;
  repeated string document_ids = 6;
  map<string, double> lateral_connections = 7;
  map<string, Edge> typed_connections = 8;
  repeated string feedforward_sources = 9;
  map<string, double> feedforward_connections = 10;
  map<string, double> feedback_connections = 11;
  double tfidf = 12;
  map<string, double> tfidf_per_doc = 13;
  double pagerank = 14;
  optional int32 cluster_id = 15;
  map<string, int32> doc_occurrence_counts = 16;
  repeated string name_tokens = 17;
}

// HierarchicalLayer represents a layer in the cortical hierarchy
message HierarchicalLayer {
  int32 level = 1;
  map<string, Minicolumn> minicolumns = 2;
}

// FloatList represents a list of floats (for embeddings)
message FloatList {
  repeated double values = 1;
}

// SemanticRelation represents a semantic relation tuple
message SemanticRelation {
  string term1 = 1;
  string relation_type = 2;
  string term2 = 3;
  double weight = 4;
}

// AnyValue represents arbitrary metadata values
message AnyValue {
  oneof value {
    string string_value = 1;
    int64 int_value = 2;
    double double_value = 3;
    bool bool_value = 4;
    AnyDict dict_value = 5;
    AnyList list_value = 6;
  }
}

// AnyDict represents arbitrary metadata dictionaries
message AnyDict {
  map<string, AnyValue> values = 1;
}

// AnyList represents arbitrary metadata lists
message AnyList {
  repeated AnyValue values = 1;
}

// ProcessorState represents the complete state of a CorticalTextProcessor
message ProcessorState {
  string version = 1;
  map<int32, HierarchicalLayer> layers = 2;
  map<string, string> documents = 3;
  map<string, AnyDict> document_metadata = 4;
  map<string, FloatList> embeddings = 5;
  repeated SemanticRelation semantic_relations = 6;
  AnyDict metadata = 7;
}
